Firmware/HomeButtonsArduino/src/static_string.h

```diff
@@ -45,16 +45,52 @@ class StaticString {
 
   size_t length() const { return strlen(data_); }
   bool empty() const { return strlen(data_) == 0; }
+  StaticString substring(size_t i) { return substring(i, length()); }
   StaticString substring(size_t i, size_t j) {
     StaticString output;
-    if (i < MAX_SIZE) {
-      auto n =
-          std::snprintf(output.data_, std::min(MAX_SIZE, j - i + 1), data_ + i);
-      _check_snprintf_return_value(n);
+    if (i > j || i >= length()) {
+      return output;
     }
+    if (j > length()) {
+      j = length();
+    }
+    auto n =
+        std::snprintf(output.data_, std::min(MAX_SIZE, j - i + 1), data_ + i);
+    _check_snprintf_return_value(n);
     return output;
   }
 
+  // removes any leading or trailing spaces
+  void trim(void) {
```

Comment:
:warning: 
Several issues in this function:
- `std::snprintf` **does not allow** overlapping input and output => either use a temporary buffer, or consider to return a new `StaticString` (the same as `substring` does). I made some test, depending on snprint implemention, result can be totally wrong.
- `if (empty())` => useless. There is always at least one null character, that is perfectly handle by the code after.
- std::min(MAX_SIZE, (size_t)(end - start) + 2) => maybe remove the `std::min`

Reply:
if `empty()` check is omitted, I think there's a problem if the string length is 0.

`length()` would return 0 and therefore `auto end = data_ + length() - 1;` would point to address just before first byte of `data_`.

The `end - start ` would then be negative causing undefined behaviour in `(size_t)(end - start)`.